src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/PtrToStringAnsiTests.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace System.Runtime.InteropServices.Tests
+{
+    public class PtrToStringAnsiTests
+    {
+        [Theory]
+        [InlineData("", 0)]
+        [InlineData("hello", 0)]
+        [InlineData("hello", 1)]
+        [InlineData("hello", 4)]
+        public void PtrToStringAnsi_Length_Success(string s, int len)
+        {
+            IntPtr ptr = Marshal.StringToCoTaskMemAnsi(s);
+            try
+            {
+                string result = Marshal.PtrToStringAnsi(ptr, len);
+                Assert.Equal(s.Substring(0, len), result);
+            }
+            finally
+            {
+                Marshal.FreeCoTaskMem(ptr);
+            }
+        }
+
+        [Fact]
+        public void PtrToStringAnsi_ZeroPointer_ReturnsNull()
+        {
+            Assert.Null(Marshal.PtrToStringAnsi(IntPtr.Zero));
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void PtrToStringAnsi_Win32AtomPointer_ReturnsNull()
```

Comment:
why the function name contains Atom? Does it relate to 64K?

Reply:
From the marshal source code there are checks:
```cs
        // Win32 has the concept of Atoms, where a pointer can either be a pointer
        // or an int.  If it's less than 64K, this is guaranteed to NOT be a 
        // pointer since the bottom 64K bytes are reserved in a process' page table.
        // We should be careful about deallocating this stuff.  Extracted to
        // a function to avoid C# problems with lack of support for IntPtr.
        // We have 2 of these methods for slightly different semantics for NULL.
        private static bool IsWin32Atom(IntPtr ptr)
        {
#if FEATURE_PAL
            return false;
#else
            long lPtr = (long)ptr;
            return 0 == (lPtr & HIWORDMASK);
#endif
        }

```

Reply:
Thanks for this.